### PR TITLE
LINK-2077 | Refund and cancel payments through webhooks

### PR DESCRIPTION
**Pull request by harriris-vincit**
### Description
- Talpa payments are no longer refunded or cancelled immediately after the requests succeed. Instead, the operations are done when the related webhook notification arrives.
- An event with paid signups cannot be cancelled or deleted. The paid signups need to be manually cancelled before the event cancellation or deletion is possible.
### Closes
[LINK-2077](https://helsinkisolutionoffice.atlassian.net/browse/LINK-2077)

[LINK-2077]: https://helsinkisolutionoffice.atlassian.net/browse/LINK-2077?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by terovirtanen** [comment]
**LINKEDEVENTS-API branch is deployed to platta: https://linkedevents-pr933.api.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
**LINKEDEVENTS-API branch is deployed to platta: https://linkedevents-pr933.api.dev.hel.ninja** :rocket::rocket::rocket:

**Comment by terovirtanen** [comment]
**LINKEDEVENTS-API branch is deployed to platta: https://linkedevents-pr933.api.dev.hel.ninja** :rocket::rocket::rocket:

File: registrations/models.py [hunk 1]
```diff
@@ -1533,36 +1537,39 @@ def partially_refund_signup_group_web_store_payment(
...
+                            signup_group
                         )
                     )
                 else:
                     payment_refunded, payment_cancelled = (
-                        self.refund_or_cancel_web_store_payment(
-                            payment,
-                            bypass_web_store_api_calls=bypass_web_store_api_calls,
-                        )
+                        self.refund_or_cancel_web_store_payment(payment)
```
Comment (nicobav) [inline registrations/models.py]: In the `refund_or_cancel_web_store_payment` there's a potential `KeyError` if the response data is wonky somehow.

File: registrations/models.py [hunk 2]
```diff
@@ -1097,26 +1121,27 @@ def delete(self, *args, **kwargs):
...
 
             try:
                 payment_refunded, payment_cancelled = (
-                    self.refund_or_cancel_web_store_payment(
-                        payment,
-                        bypass_web_store_api_calls=kwargs.pop(
-                            "bypass_web_store_api_calls", False
-                        ),
-                    )
+                    self.refund_or_cancel_web_store_payment(payment)
```
Comment (nicobav) [inline registrations/models.py]: Potential KeyError can be raised in the call loop.

File: registrations/models.py [hunk 3]
```diff
@@ -946,6 +946,40 @@ def create_web_store_payment(self):
...
+
+        if isinstance(self, SignUp):
+            signup_relation_kwarg = {"signup": self}
+        else:
+            signup_relation_kwarg = {"signup_group": self}
+        SignUpPaymentRefund.objects.create(
+            **signup_relation_kwarg,
+            payment=payment,
+            amount=self.total_payment_amount,
+            external_refund_id=resp_json["refunds"][0]["refundId"],
```
Comment (nicobav) [inline registrations/models.py:969]: there's a potential KeyError or `IndexError` if the response data is wonky somehow. 
What I understood this is part of call loop in deletion and it does not catch `KeyError` nor `IndexError`

Comment (harriris-vincit) [reply registrations/models.py]: Added logic to catch this, log it with `logger` and raise a `WebStoreRefundValidationError` to aid quite a bit with debugging.

**Review by nicobav — APPROVED** [review]
👍

**Comment by terovirtanen** [comment]
**LINKEDEVENTS-API branch is deployed to platta: https://linkedevents-pr933.api.dev.hel.ninja** :rocket::rocket::rocket:

**Pull request merged**